From e02bc29b2cfa7806830d6da8b2322cddd67e8dfe Mon Sep 17 00:00:00 2001
From: Paul Kocialkowski <paul.kocialkowski@bootlin.com>
Date: Wed, 24 Apr 2019 11:04:13 +0200
Subject: drm/sun4i: Unbind components before releasing DRM and memory
Git-commit: e02bc29b2cfa7806830d6da8b2322cddd67e8dfe
Patch-mainline: v5.1-rc7
References: bsc#1113722

Our components may still be using the DRM device driver (if only to
access our driver's private data), so make sure to unbind them before
the final drm_dev_put.

Also release our reserved memory after component unbind instead of
before to match reverse creation order.

Fixes: f5a9ed867c83 ("drm/sun4i: Fix component unbinding and component master deletion")
Signed-off-by: Paul Kocialkowski <paul.kocialkowski@bootlin.com>
Reviewed-by: Chen-Yu Tsai <wens@csie.org>
Link: https://patchwork.freedesktop.org/patch/msgid/20190424090413.6918-1-paul.kocialkowski@bootlin.com
Acked-by: Thomas Zimmermann <tzimmermann@suse.de>
---
 drivers/gpu/drm/sun4i/sun4i_drv.c |    5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

--- a/drivers/gpu/drm/sun4i/sun4i_drv.c
+++ b/drivers/gpu/drm/sun4i/sun4i_drv.c
@@ -167,10 +167,11 @@ static void sun4i_drv_unbind(struct devi
 	drm_atomic_helper_shutdown(drm);
 	drm_mode_config_cleanup(drm);
 	drm_vblank_cleanup(drm);
-	of_reserved_mem_device_release(dev);
-	drm_dev_unref(drm);
 
 	component_unbind_all(dev, NULL);
+	of_reserved_mem_device_release(dev);
+
+	drm_dev_unref(drm);
 }
 
 static const struct component_master_ops sun4i_drv_master_ops = {
